e2e/tests/wasm/grandpa_test.go

```diff
@@ -255,6 +257,160 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
 	s.Require().True(parachainUserStake.Amount.Equal(math.NewInt(amountToSend-amountToReflect)), "parachain user's final stake amount not expected")
 }
 
+func (s *GrandpaTestSuite) TestRecoverClient_Succeeds_GrandpaContract() {
+	ctx := context.Background()
+
+	// set the trusting period to a value which will still be valid upon client creation, but invalid before the first update
+	var badTrustingPeriod = time.Second * 10
+
+	chainA, chainB := s.GetChains(func(options *testsuite.ChainOptions) {
+		// configure chain A (polkadot)
+		options.ChainASpec.ChainName = composable
+		options.ChainASpec.Type = "polkadot"
+		options.ChainASpec.ChainID = "rococo-local"
+		options.ChainASpec.Name = "composable"
+		options.ChainASpec.Images = []ibc.DockerImage{
+			// TODO: https://github.com/cosmos/ibc-go/issues/4965
+			{
+				Repository: "ghcr.io/misko9/polkadot-node",
+				Version:    "local",
+				UidGid:     "1000:1000",
+			},
+			{
+				Repository: "ghcr.io/misko9/parachain-node",
+				Version:    "latest",
+				UidGid:     "1000:1000",
+			},
+		}
+		options.ChainASpec.Bin = "polkadot"
+		options.ChainASpec.Bech32Prefix = composable
+		options.ChainASpec.Denom = "uDOT"
+		options.ChainASpec.GasPrices = ""
+		options.ChainASpec.GasAdjustment = 0
+		options.ChainASpec.TrustingPeriod = ""
+		options.ChainASpec.CoinType = "354"
+
+		// these values are set by default for our cosmos chains, we need to explicitly remove them here.
+		options.ChainASpec.ModifyGenesis = nil
+		options.ChainASpec.ConfigFileOverrides = nil
+		options.ChainASpec.EncodingConfig = nil
+
+		// configure chain B (cosmos)
+		options.ChainBSpec.ChainName = simd // Set chain name so that a suffix with a "dash" is not appended (required for hyperspace)
+		options.ChainBSpec.Type = "cosmos"
+		options.ChainBSpec.Name = "simd"
+		options.ChainBSpec.ChainID = simd
+		options.ChainBSpec.Bin = simd
+		options.ChainBSpec.Bech32Prefix = "cosmos"
+
+		// TODO: hyperspace relayer assumes a denom of "stake", hard code this here right now.
+		// https://github.com/cosmos/ibc-go/issues/4964
+		options.ChainBSpec.Denom = "stake"
+		options.ChainBSpec.GasPrices = "0.00stake"
+		options.ChainBSpec.GasAdjustment = 1
+		options.ChainBSpec.TrustingPeriod = "504h"
+		options.ChainBSpec.CoinType = "118"
+
+		options.ChainBSpec.ChainConfig.NoHostMount = false
+		options.ChainBSpec.ConfigFileOverrides = getConfigOverrides()
+		options.ChainBSpec.EncodingConfig = testsuite.SDKEncodingConfig()
+	})
+
+	polkadotChain := chainA.(*polkadot.PolkadotChain)
+	cosmosChain := chainB.(*cosmos.CosmosChain)
+
+	// we explicitly skip path creation as the contract needs to be uploaded before we can create clients.
+	r := s.ConfigureRelayer(ctx, polkadotChain, cosmosChain, nil, func(options *interchaintest.InterchainBuildOptions) {
+		options.SkipPathCreation = true
+	})
+
+	s.InitGRPCClients(cosmosChain)
```

Comment:
I think all of this code can probably put into a setup fn for the `GrandpaTestSuite` type to be re-used in all the tests.

Reply:
I 100% agree. I can do it as part of this PR, or separate it out

Reply:
resolved in https://github.com/cosmos/ibc-go/pull/5027/commits/dcc1a11ef6607507b39ba73ae1f3c8b2054d9515